File: bench/maps.js
// bun maps.js — mirror of maps.b / maps.go (hash map throughput)
const n = Number(process.argv[2] ?? 400000);
const seed = Number(process.argv[3] ?? 1);
const m = new Map();
for (let i = 0; i < n; i++) m.set(i + seed, i * 2 + seed);

let sum = 0;
for (let i = 0; i < n; i++) sum += m.get(i + seed);

let hits = 0;
for (let i = 0; i < n; i++) {
    if (m.has(i + seed)) hits++;
    if (m.has(i + n + seed)) hits++;
}

const sn = Math.floor(n / 5);
const sm = new Map();
for (let i = 0; i < sn; i++) sm.set(`key${seed}_${i}`, i + seed);
let ssum = 0;
for (let i = 0; i < sn; i++) ssum += sm.get(`key${seed}_${i}`);

console.log(`maps ${sum} ${hits} ${ssum} ${m.size} ${sm.size}`);

File: bench/maps.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func arg(at, fallback int) int {
	if len(os.Args) <= at+1 {
		return fallback
	}
	n, err := strconv.Atoi(os.Args[at+1])
	if err != nil {
		return fallback
	}
	return n
}

func main() {
	n, seed := arg(0, 400000), arg(1, 1)
	m := make(map[int]int, n) // size is known: preallocate, the Go idiom
	for i := 0; i < n; i++ {
		m[i+seed] = i*2 + seed
	}

	sum := 0
	for i := 0; i < n; i++ {
		sum += m[i+seed]
	}

	hits := 0
	for i := 0; i < n; i++ {
		if _, ok := m[i+seed]; ok {
			hits++
		}
		if _, ok := m[i+n+seed]; ok {
			hits++
		}
	}

	sn := n / 5
	sm := make(map[string]int, sn)
	for i := 0; i < sn; i++ {
		sm["key"+strconv.Itoa(seed)+"_"+strconv.Itoa(i)] = i + seed
	}
	ssum := 0
	kb := make([]byte, 0, 16)
	for i := 0; i < sn; i++ {
		kb = append(kb[:0], "key"...)
		kb = strconv.AppendInt(kb, int64(seed), 10)
		kb = append(kb, '_')
		kb = strconv.AppendInt(kb, int64(i), 10)
		ssum += sm[string(kb)] // the compiler elides this alloc for map reads
	}

	fmt.Printf("maps %d %d %d %d %d\n", sum, hits, ssum, len(m), len(sm))
}
